File: backend/internal/crdttest/crdt_convergence_test.go
package crdttest

import (
	"fmt"
	"math/rand/v2"
	"os"
	"testing"

	"github.com/reearth/ygo/crdt"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// IMPORTANT: ygo requires GetText/GetMap to be called OUTSIDE Transact.
// Calling doc.GetText() inside a Transact closure causes a deadlock because
// GetText acquires a read lock that Transact already holds as a write lock.

// =============================================================================
// Teste 1 — Go sozinho, sem rede
// =============================================================================

func TestCRDT1_GoConvergence(t *testing.T) {
	t.Run("fixed_concurrent_inserts", func(t *testing.T) {
		docA := crdt.New(crdt.WithGC(false))
		docB := crdt.New(crdt.WithGC(false))

		// Pre-register YText outside of Transact
		textA := docA.GetText("content")
		textB := docB.GetText("content")

		const initial = "hello world"
		docA.Transact(func(txn *crdt.Transaction) {
			textA.Insert(txn, 0, initial, nil)
		})
		stateA := docA.EncodeStateAsUpdate()
		require.NoError(t, docB.ApplyUpdate(stateA))

		require.Equal(t, initial, textA.ToString())
		require.Equal(t, initial, textB.ToString())

		// Edições concorrentes — ANTES de trocar updates
		docA.Transact(func(txn *crdt.Transaction) {
			textA.Insert(txn, 5, "XXX", nil)
		})
		docB.Transact(func(txn *crdt.Transaction) {
			textB.Insert(txn, 3, "YYY", nil)
		})

		// Troca de updates
		updateA := docA.EncodeStateAsUpdate()
		updateB := docB.EncodeStateAsUpdate()
		require.NoError(t, docA.ApplyUpdate(updateB))
		require.NoError(t, docB.ApplyUpdate(updateA))

		resultA := textA.ToString()
		resultB := textB.ToString()

		assert.Equal(t, resultA, resultB,
			"CRDT divergiu:\n  docA=%q\n  docB=%q", resultA, resultB)
		assert.Equal(t, 1, crdtCount(resultA, "XXX"), "XXX duplicado: %q", resultA)
		assert.Equal(t, 1, crdtCount(resultA, "YYY"), "YYY duplicado: %q", resultA)
		t.Logf("✅ Resultado convergido: %q", resultA)
	})

	t.Run("fuzz_30_concurrent_inserts", func(t *testing.T) {
		const iterations = 30
		for i := range iterations {
			docA := crdt.New(crdt.WithGC(false))
			docB := crdt.New(crdt.WithGC(false))

			textA := docA.GetText("content")
			textB := docB.GetText("content")

			const initial = "abcdefghijklmnopqrstuvwxyz"
			docA.Transact(func(txn *crdt.Transaction) {
				textA.Insert(txn, 0, initial, nil)
			})
			stateA := docA.EncodeStateAsUpdate()
			require.NoError(t, docB.ApplyUpdate(stateA), "iter %d: sync inicial", i)

			length := textA.Len()
			posA := rand.IntN(length + 1)
			posB := rand.IntN(length + 1)

			docA.Transact(func(txn *crdt.Transaction) {
				textA.Insert(txn, posA, "XXX", nil)
			})
			docB.Transact(func(txn *crdt.Transaction) {
				textB.Insert(txn, posB, "YYY", nil)
			})

			updateA := docA.EncodeStateAsUpdate()
			updateB := docB.EncodeStateAsUpdate()
			require.NoError(t, docA.ApplyUpdate(updateB), "iter %d: docA←B", i)
			require.NoError(t, docB.ApplyUpdate(updateA), "iter %d: docB←A", i)

			resultA := textA.ToString()
			resultB := textB.ToString()

			if !assert.Equal(t, resultA, resultB,
				"iter %d: CRDT divergiu posA=%d posB=%d\n  docA=%q\n  docB=%q",
				i, posA, posB, resultA, resultB) {
				t.FailNow()
			}
			assert.Equal(t, 1, crdtCount(resultA, "XXX"), "iter %d: XXX duplicado", i)
			assert.Equal(t, 1, crdtCount(resultA, "YYY"), "iter %d: YYY duplicado", i)
		}
		t.Logf("✅ %d iterações de fuzzing passaram", iterations)
	})
}

// =============================================================================
// Teste 3 (lado Go) — Fixture binário para interop com Dart
// =============================================================================

func TestCRDT3_GenerateGoFixture(t *testing.T) {
	docA := crdt.New(crdt.WithGC(false))
	textA := docA.GetText("content")

	docA.Transact(func(txn *crdt.Transaction) {
		textA.Insert(txn, 0, "hello world", nil)
	})
	docA.Transact(func(txn *crdt.Transaction) {
		textA.Insert(txn, 5, "GO_EDIT", nil)
	})

	update := docA.EncodeStateAsUpdate()

	// Self-check
	docB := crdt.New(crdt.WithGC(false))
	textB := docB.GetText("content")
	require.NoError(t, docB.ApplyUpdate(update))
	const expected = "helloGO_EDIT world"
	assert.Equal(t, expected, textB.ToString())

	require.NoError(t, crdtWriteFixture("testdata/crdt3_go_update.bin", update))
	require.NoError(t, crdtWriteFixture("testdata/crdt3_go_expected.txt", []byte(expected)))
	t.Logf("✅ Fixture Go gerado: %d bytes, texto: %q", len(update), expected)
}

func TestCRDT3_ApplyDartUpdate(t *testing.T) {
	data, err := os.ReadFile("testdata/crdt3_dart_update.bin")
	if err != nil {
		t.Skipf("Fixture Dart não encontrado — rode o Teste 2 Dart primeiro: %v", err)
	}
	expected, err := os.ReadFile("testdata/crdt3_dart_expected.txt")
	if err != nil {
		t.Skipf("Expected text não encontrado: %v", err)
	}

	doc := crdt.New(crdt.WithGC(false))
	doc.GetText("content")
	require.NoError(t, doc.ApplyUpdate(data), "Go falhou ao aplicar update Dart")

	// Re-read after apply (GetText outside transact is fine for read)
	got := doc.GetText("content").ToString()
	assert.Equal(t, string(expected), got,
		"Go não reconheceu o update Dart:\n  expected=%q\n  got=%q", string(expected), got)
	t.Logf("✅ Update Dart aplicado no Go: %q", got)
}

// =============================================================================
// helpers
// =============================================================================

func crdtCount(s, sub string) int {
	count := 0
	for i := 0; i+len(sub) <= len(s); i++ {
		if s[i:i+len(sub)] == sub {
			count++
			i += len(sub) - 1
		}
	}
	return count
}

func crdtWriteFixture(path string, data []byte) error {
	if err := os.MkdirAll("testdata", 0o755); err != nil {
		return fmt.Errorf("mkdir testdata: %w", err)
	}
	return os.WriteFile(path, data, 0o644)
}

File: test/crdt_validation/crdt_convergence_test.dart
import 'dart:io';
import 'dart:math';
import 'dart:typed_data';

import 'package:flutter_test/flutter_test.dart';
import 'package:dart_crdt/dart_crdt.dart';

SharedType _text(Doc doc, String key) {
  return doc.getText(key);
}

String _str(Doc doc, String key) => doc.getText(key).toPlainText();

// =============================================================================
// Teste 2 — Dart sozinho, sem rede
// Verifica que dois Docs yjs_dart convergem após trocarem updates concorrentes.
// =============================================================================

void main() {
  group('Teste 2 — yjs_dart convergência (sem rede)', () {
    test('inserts concorrentes fixos convergem', () {
      final docA = Doc();
      final docB = Doc();

      // Pré-registro do tipo (necessário para applyUpdate funcionar corretamente)
      docA.getText('content');
      docB.getText('content');

      const initial = 'hello world';

      // Ambos partem do mesmo estado inicial
      docA.transact((txn) {
        _text(docA, 'content').insertText(0, initial);
      });
      // Sync initial state: send full state of A to B
      final stateA = encodeStateAsUpdate(docA);
      applyUpdate(docB, stateA);

      expect(_str(docA, 'content'), initial);
      expect(_str(docB, 'content'), initial);

      // Edições concorrentes — antes de trocar updates
      // Capture state vectors BEFORE the concurrent edits
      final svA = encodeDocumentStateVector(docA);
      final svB = encodeDocumentStateVector(docB);

      docA.transact((txn) {
        _text(docA, 'content').insertText(5, 'XXX');
      });
      docB.transact((txn) {
        _text(docB, 'content').insertText(3, 'YYY');
      });

      // Troca de updates — only the DIFF since svA/svB (not full state)
      final diffA = encodeStateAsUpdate(docA, svB); // what B is missing
      final diffB = encodeStateAsUpdate(docB, svA); // what A is missing
      applyUpdate(docA, diffB);
      applyUpdate(docB, diffA);

      final textA = _str(docA, 'content');
      final textB = _str(docB, 'content');

      // ✅ Critério 1: convergência
      expect(textA, equals(textB),
          reason: 'CRDT divergiu:\n  docA=$textA\n  docB=$textB');
      // ✅ Critério 2: sem duplicatas
      expect(_count(textA, 'XXX'), 1, reason: 'XXX duplicado em: $textA');
      expect(_count(textA, 'YYY'), 1, reason: 'YYY duplicado em: $textA');

      // ignore: avoid_print
      print('✅ Resultado convergido: "$textA"');
    });

    test('fuzzing — 30 inserts concorrentes em posições aleatórias', () {
      const iterations = 30;
      final rng = Random(42); // seed fixo para reprodutibilidade

      for (var i = 0; i < iterations; i++) {
        final docA = Doc();
        final docB = Doc();
        docA.getText('content');
        docB.getText('content');

        const initial = 'abcdefghijklmnopqrstuvwxyz';
      docA.transact((txn) {
        _text(docA, 'content').insertText(0, initial);
      });
      final stateA = encodeStateAsUpdate(docA);
      applyUpdate(docB, stateA);

      final len = _text(docA, 'content').toPlainText().length;
      final posA = rng.nextInt(len + 1);
      final posB = rng.nextInt(len + 1);

      // Capture state vectors BEFORE the concurrent edits
      final svA = encodeDocumentStateVector(docA);
      final svB = encodeDocumentStateVector(docB);

      docA.transact((txn) {
        _text(docA, 'content').insertText(posA, 'XXX');
      });
      docB.transact((txn) {
        _text(docB, 'content').insertText(posB, 'YYY');
      });

        // Exchange only the DIFF (not full state)
        final diffA = encodeStateAsUpdate(docA, svB);
        final diffB = encodeStateAsUpdate(docB, svA);
        applyUpdate(docA, diffB);
        applyUpdate(docB, diffA);

        final textA = _str(docA, 'content');
        final textB = _str(docB, 'content');

        expect(textA, equals(textB),
            reason:
                'iter $i: CRDT divergiu posA=$posA posB=$posB\n  docA=$textA\n  docB=$textB');
        expect(_count(textA, 'XXX'), 1,
            reason: 'iter $i: XXX duplicado em $textA');
        expect(_count(textA, 'YYY'), 1,
            reason: 'iter $i: YYY duplicado em $textA');
      }
      // ignore: avoid_print
      print('✅ $iterations iterações de fuzzing passaram');
    });
  });

  // ===========================================================================
  // Teste 3 (lado Dart) — Geração de fixture + interop com Go
  // ===========================================================================

  group('Teste 3 — interop binário Go ↔ Dart', () {
    test('Dart gera fixture para o Go aplicar', () {
      final docA = Doc();
      docA.getText('content');

      docA.transact((txn) {
        _text(docA, 'content').insertText(0, 'hello world');
      });
      // Edição conhecida: insere "DART_EDIT" na posição 5
      docA.transact((txn) {
        _text(docA, 'content').insertText(5, 'DART_EDIT');
      });

      final update = encodeStateAsUpdate(docA);

      // Self-check: outro doc Dart consegue aplicar?
      final docB = Doc();
      docB.getText('content');
      applyUpdate(docB, update);

      final got = _str(docB, 'content');
      const expected = 'helloDARTEDIT world';
      // Nota: "DART_EDIT" sem underscore → "DART_EDIT" com underscore → depende da lib
      // Verificamos o que realmente sai
      expect(got.contains('DART_EDIT'), isTrue,
          reason: 'DART_EDIT ausente: $got');

      final dir = Directory('testdata');
      if (!dir.existsSync()) dir.createSync(recursive: true);
      File('testdata/crdt3_dart_update.bin').writeAsBytesSync(update);
      File('testdata/crdt3_dart_expected.txt').writeAsStringSync(got);

      // ignore: avoid_print
      print('✅ Fixture Dart gerado: ${update.length} bytes, texto: "$got"');
    });

    test('Dart aplica fixture gerado pelo Go', () {
      const fixturePath = 'testdata/crdt3_go_update.bin';
      const expectedPath = 'testdata/crdt3_go_expected.txt';

      final fixtureFile = File(fixturePath);
      if (!fixtureFile.existsSync()) {
        // ignore: avoid_print
        print(
            'SKIP: fixture Go não encontrado em $fixturePath — rode TestCRDT3_GenerateGoFixture primeiro');
        return;
      }

      final update = Uint8List.fromList(fixtureFile.readAsBytesSync());
      final expected = File(expectedPath).readAsStringSync().trim();

      final doc = Doc();
      doc.getText('content');
      applyUpdate(doc, update);

      final got = _str(doc, 'content');
      expect(got, equals(expected),
          reason: 'Dart não reconheceu o update Go:\n  expected=$expected\n  got=$got');
      // ignore: avoid_print
      print('✅ Update Go aplicado no Dart: "$got"');
    });

    test('edição concorrente cross-lib converge (Dart simula dois peers)', () {
      const fixturePath = 'testdata/crdt3_go_update.bin';
      final fixtureFile = File(fixturePath);
      if (!fixtureFile.existsSync()) {
        // ignore: avoid_print
        print('SKIP: fixture Go ausente — rode TestCRDT3_GenerateGoFixture');
        return;
      }

      final goUpdate = Uint8List.fromList(fixtureFile.readAsBytesSync());

      // peerA simula o que o Go fez: parte do estado vazio, aplica o update do Go
      final peerA = Doc();
      peerA.getText('content');
      applyUpdate(peerA, goUpdate);

      // peerB começa do mesmo estado base (sem o update do Go)
      // e faz uma edição local diferente
      final peerB = Doc();
      peerB.getText('content');
      peerB.transact((txn) {
        _text(peerB, 'content').insertText(0, 'hello world');
      });
      peerB.transact((txn) {
        _text(peerB, 'content').insertText(0, 'DART_PREFIX_');
      });

      // Agora troca updates
      final updateA = encodeStateAsUpdate(peerA);
      final updateB = encodeStateAsUpdate(peerB);
      applyUpdate(peerA, updateB);
      applyUpdate(peerB, updateA);

      final textA = _str(peerA, 'content');
      final textB = _str(peerB, 'content');

      expect(textA, equals(textB),
          reason: 'Cross-lib CRDT divergiu:\n  peerA=$textA\n  peerB=$textB');
      expect(_count(textA, 'GO_EDIT'), 1,
          reason: 'GO_EDIT duplicado: $textA');
      expect(_count(textA, 'DART_PREFIX_'), 1,
          reason: 'DART_PREFIX_ duplicado: $textA');
      // ignore: avoid_print
      print('✅ Convergência cross-lib: "$textA"');
    });
  });
}

// =============================================================================
// helpers
// =============================================================================

int _count(String s, String sub) {
  int count = 0;
  int start = 0;
  while (true) {
    final idx = s.indexOf(sub, start);
    if (idx == -1) break;
    count++;
    start = idx + sub.length;
  }
  return count;
}
